IO-649 - Improve the performance of the contentEquals() methods.

brettlounsbury opened the pull request:
This change modifies the contentEquals() methods to internally buffer content into a byte/char array and to then do batch comparisons of those arrays using Arrays.equals instead of using a BufferedInputStream or BufferedReader and making use of the single byte/char read() methods.

This reduces the number of method invocations by a factor equal to the buffer size and avoids casting every byte read to an int and improves performance significantly.

The following table shows the performance increase over 1000 iterations of comparing 2 1GB InputStream of binary data (stored in memory to avoid I/O).  This test was performed on an EC2 M4.4XL host using Java 1.8.0.232 and there was a forced System.gc() between each iteration to avoid GC as a source of latency:

Average: 7236 to 858ms (8.43x speedup)
P50: 7224 to 856ms (8.44x speedup)
P90: 7249 to 860ms (8.43x speedup)
P99: 7410 to 913ms (8.12x speedup)
P100: 8330 to 1278ms (6.52x speedup)

The following table shows the performance increase over 1000 iterations of comparing 2 1GB Reader of character data (stored in memory to avoid I/O).  This test was performed on an EC2 M4.4XL host using Java 1.8.0.232 and there was a forced System.gc() between each iteration to avoid GC as a source of latency:

Average: 11281 to 1737ms (6.50x speedup)
P50: 11262 to 1735ms (6.49x speedup)
P90: 11292 to 1741ms (6.49x speedup)
P99: 11707 to 1774ms (6.60x speedup)
P100: 12176 to 1884ms (6.46x speedup)

brettlounsbury commented:
https://issues.apache.org/jira/browse/IO-649

michael-o commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -708,31 +706,79 @@ public static void closeQuietly(final Writer output) {
+    public static boolean contentEquals(final InputStream input1, final InputStream input2, final int bufferSize)
+            throws IOException {
+        if (bufferSize <= 0) {
+            throw new IllegalArgumentException("Buffer size must be positive: " + bufferSize);
```
This one is not documented.

brettlounsbury replied:
Fixed and updated.  Done for both methods that throw an IllegalArgumentException.

michael-o commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -708,31 +706,79 @@ public static void closeQuietly(final Writer output) {
+     * @param bufferSize the size of the internal buffer to use.
+     * @return true if the content of the streams are equal or they both don't
+     * exist, false otherwise
+     * @throws NullPointerException if either input is null
```
I fail to see where this one is thrown.

brettlounsbury replied:
This documentation is a clone/mutate from the original method's javadoc (the one without the bufferSize).  This doesn't look to be explicitly thrown, but if either InputStream is null it throws due to the stream being dereferenced.

Check the javadoc for the `public static boolean contentEquals(final InputStream input1, final InputStream input2)` method.

Happy to either (a) explicitly throw a NullPointerException up front, (b) remove the comment from both places, or (c) do nothing.

michael-o replied:
I would rather see an explicit NPE up front.

brettlounsbury replied:
I looked at the beginning of each method again.  Both contentEquals methods start with the following statements.  This can never throw a NullPointerException.  If both streams are null they will be equal and therefore always return true.  If one stream is null and the other is not it will always return false based on the XOR logic.  Only if both streams are non-null and not the same object will the actual logic of the method execute.

```
if (input1 == input2) {
            return true;
}
if (input1 == null ^ input2 == null) {
            return false;
}
```

michael-o commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -746,25 +792,74 @@ public static boolean contentEquals(final InputStream input1, final InputStream
+     * @param bufferSize the size of the internal buffer to use.
+     * @return true if the content of the readers are equal or they both don't
+     * exist, false otherwise
+     * @throws NullPointerException if either input is null
```
Same here.

brettlounsbury replied:
Same answer as above :-)

brettlounsbury commented:
I took a look at the coverage drop.  I forgot to add tests that exercise the bounds checking of the bufferSize inputs to both contentEquals method.  I've pushed a new version with that.

brettlounsbury commented:
@garydgregory - can you take a look or propose another reviewer?

garydgregory commented:
Hi @brettlounsbury.

Thank you for your PR and updates. I've sprinkled comments throughout.

Performance claims must be backed up somehow, please add a JMH benchmark to the test so myself and others can compare mileage. 

For examples of JMH usage, please see other Apache Commons components like Commons BCEL, Commons Crypto, Commons CSV, Commons Lang, Commons Math, and Commons RNG.

Thank you!
Gary

brettlounsbury commented:
Hi @garydgregory,

I don't see any comments from you other than the request for a benchmark. 

I have updated IOUtils to include a test dependency on JMH and and to include a benchmark that shows BufferedInputStream.read() performance compared to InputStream.read(byte[]) performance.  This test creates an 8MB file on disk once before the benchmark is ran and reads it back in.  It then just keeps a running sum of the bytes in the file (and ignores any overflow/underflow) and blackhole's the sum at the end to avoid any compiler optimization.

Results look something like this:
Benchmark                                                        Mode  Cnt         Score        Error  Units
SingleByteReadVsMultiByteReadPerformanceTest.testReadMultiBytes  avgt   25   6220618.290 ± 204562.812  ns/op
SingleByteReadVsMultiByteReadPerformanceTest.testReadSingleByte  avgt   25  21740409.146 ± 711294.924  ns/op

brettlounsbury commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
+     * fully consumed.
+     * @throws IOException if an I/O error occurs.
+     */
+    private static int fillBuffer(final InputStream inputStream, final byte[] buffer) throws IOException {
```
Yes, it looks like IOUtils.read(InputStream, byte[]) and IOUtils.read(Reader, char[]) (without the optional arguments) should do exactly what I want.

Thanks for the call out, I didn't realize those methods existed.

brettlounsbury commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
+    }
+
+    /**
+     * Compares the contents of two Streams to determine if they are equal or not.
```
Done.

brettlounsbury commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
+     * exist, false otherwise
+     * @throws IllegalArgumentException if bufferSize is less than or equal to zero.
+     * @throws IOException              if an I/O error occurs
+     * @since 2.6
```
Done

brettlounsbury commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
+        do {
+            totalBytesRead += bytesRead;
+            bytesRead = inputStream.read(buffer, totalBytesRead, buffer.length - totalBytesRead);
+        } while (bytesRead != -1 && totalBytesRead < buffer.length);
```
Good catch.  I've used it elsewhere but somehow missed it in the loops.  Its somewhat moot, however, since the two fillBuffer methods have been deleted.

brettlounsbury commented on src/main/java/org/apache/commons/io/IOUtils.java:
```diff
@@ -691,80 +692,173 @@ public static void closeQuietly(final Writer output) {
+        do {
+            totalBytesRead += bytesRead;
+            bytesRead = reader.read(buffer, totalBytesRead, buffer.length - totalBytesRead);
+        } while (bytesRead != -1 && totalBytesRead < buffer.length);
```
Same as above :-)

brettlounsbury commented:
@garydgregory,

I really appreciate your thorough review of my code.  I think at this point I have addressed your comments, but if you have any others please let me know.

I will add that I made the IOUtils.DEFAULT_BUFFER_SIZE constant package private (default visibility) so it could be referenced from IOUtils tests so that I can ensure the logic makes sense around buffer boundaries and I didn't want to clone that property since it could later be mutated and invalidate some of the edge condition testing being done.

brettlounsbury commented:
@garydgregory - Any other feedback here or anything else I can do to improve this?

brettlounsbury commented:
@garydgregory - Can you take another look or suggest another reviewer?

garydgregory commented:
Sorry about the delay, last day of vacation here ;-)

brettlounsbury commented:
> Sorry about the delay, last day of vacation here ;-)

No worries - hope you had a great vacation!

brettlounsbury commented:
@garydgregory - can you take another look when you have a chance.

brettlounsbury commented:
@garydgregory - I think all necessary changes have been made.  Are you able to take a look?

garydgregory commented:
@brettlounsbury May you please rebase on master?

brettlounsbury commented:
@garydgregory - resolved the merge conflict, should be good now.

garydgregory commented:
Closing, no feedback.
